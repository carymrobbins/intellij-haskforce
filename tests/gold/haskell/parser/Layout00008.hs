module Layout00008 where

-- | Calculate the normal form CRC using a table
crcNormal :: (Bits a)
          => Pull1 a -> Data a -> Pull1 Word8 -> Data a
crcNormal table initial xs = fromZero $ fold step initial xs
  where
    sz         = bitSize initial
    step crc a = (table ! (Z :. i2n ((i2n (crc .>>. (sz - 8)) .&. 0xFF) `xor` a))) `xor` (crc .<<. 8)
