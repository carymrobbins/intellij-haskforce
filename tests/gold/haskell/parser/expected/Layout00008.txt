Haskell File
  MODULEDECL
    PsiElement(HaskellTokenType.module)('module')
    QCONID
      CONID
        PsiElement(HaskellTokenType.conidRegexp)('Layout00008')
    PsiElement(HaskellTokenType.where)('where')
  PsiComment(HaskellTokenType.haddock)('-- | Calculate the normal form CRC using a table\n')
  BODY
    GENDECL
      VARS
        VARID
          PsiElement(HaskellTokenType.varidRegexp)('crcNormal')
      PsiElement(HaskellTokenType.::)('::')
      CTYPE
        CLSCONTEXT
          ATYPE
            PsiElement(HaskellTokenType.()('(')
            CTYPE
              TYPEE
                ATYPE
                  OQTYCON
                    QTYCON
                      TYCON
                        CONID
                          PsiElement(HaskellTokenType.conidRegexp)('Bits')
                ATYPE
                  TYVAR
                    VARID
                      PsiElement(HaskellTokenType.varidRegexp)('a')
            PsiElement(HaskellTokenType.))(')')
        PsiElement(HaskellTokenType.=>)('=>')
        CTYPE
          TYPEE
            ATYPE
              OQTYCON
                QTYCON
                  TYCON
                    CONID
                      PsiElement(HaskellTokenType.conidRegexp)('Pull1')
            ATYPE
              TYVAR
                VARID
                  PsiElement(HaskellTokenType.varidRegexp)('a')
            PsiElement(HaskellTokenType.->)('->')
            TYPEE
              ATYPE
                OQTYCON
                  QTYCON
                    TYCON
                      CONID
                        PsiElement(HaskellTokenType.conidRegexp)('Data')
              ATYPE
                TYVAR
                  VARID
                    PsiElement(HaskellTokenType.varidRegexp)('a')
              PsiElement(HaskellTokenType.->)('->')
              TYPEE
                ATYPE
                  OQTYCON
                    QTYCON
                      TYCON
                        CONID
                          PsiElement(HaskellTokenType.conidRegexp)('Pull1')
                ATYPE
                  OQTYCON
                    QTYCON
                      TYCON
                        CONID
                          PsiElement(HaskellTokenType.conidRegexp)('Word8')
                PsiElement(HaskellTokenType.->)('->')
                TYPEE
                  ATYPE
                    OQTYCON
                      QTYCON
                        TYCON
                          CONID
                            PsiElement(HaskellTokenType.conidRegexp)('Data')
                  ATYPE
                    TYVAR
                      VARID
                        PsiElement(HaskellTokenType.varidRegexp)('a')
    FUNORPATDECL
      VARID
        PsiElement(HaskellTokenType.varidRegexp)('crcNormal')
      VARID
        PsiElement(HaskellTokenType.varidRegexp)('table')
      VARID
        PsiElement(HaskellTokenType.varidRegexp)('initial')
      VARID
        PsiElement(HaskellTokenType.varidRegexp)('xs')
      RHS
        PsiElement(HaskellTokenType.=)('=')
        EXP
          QVAR
            QVARID
              VARID
                PsiElement(HaskellTokenType.varidRegexp)('fromZero')
          QOP
            QVAROP
              QVARSYM
                VARSYM
                  PsiElement(HaskellTokenType.$)('$')
          QVAR
            QVARID
              VARID
                PsiElement(HaskellTokenType.varidRegexp)('fold')
          QVAR
            QVARID
              VARID
                PsiElement(HaskellTokenType.varidRegexp)('step')
          QVAR
            QVARID
              VARID
                PsiElement(HaskellTokenType.varidRegexp)('initial')
          QVAR
            QVARID
              VARID
                PsiElement(HaskellTokenType.varidRegexp)('xs')
        PsiElement(HaskellTokenType.where)('where')
        FUNORPATDECL
          PAT
            VARID
              PsiElement(HaskellTokenType.varidRegexp)('sz')
          RHS
            PsiElement(HaskellTokenType.=)('=')
            EXP
              QVAR
                QVARID
                  VARID
                    PsiElement(HaskellTokenType.varidRegexp)('bitSize')
              QVAR
                QVARID
                  VARID
                    PsiElement(HaskellTokenType.varidRegexp)('initial')
        FUNORPATDECL
          VARID
            PsiElement(HaskellTokenType.varidRegexp)('step')
          VARID
            PsiElement(HaskellTokenType.varidRegexp)('crc')
          VARID
            PsiElement(HaskellTokenType.varidRegexp)('a')
          RHS
            PsiElement(HaskellTokenType.=)('=')
            EXP
              PsiElement(HaskellTokenType.()('(')
              EXP
                QVAR
                  QVARID
                    VARID
                      PsiElement(HaskellTokenType.varidRegexp)('table')
                QOP
                  QVAROP
                    QVARSYM
                      VARSYM
                        PsiElement(HaskellTokenType.!)('!')
                PsiElement(HaskellTokenType.()('(')
                EXP
                  QCON
                    QCONID
                      CONID
                        PsiElement(HaskellTokenType.conidRegexp)('Z')
                  QOP
                    QCONOP
                      GCONSYM
                        QCONSYM
                          CONSYM
                            PsiElement(HaskellTokenType.Consym)(':.')
                  QVAR
                    QVARID
                      VARID
                        PsiElement(HaskellTokenType.varidRegexp)('i2n')
                  PsiElement(HaskellTokenType.()('(')
                  EXP
                    PsiElement(HaskellTokenType.()('(')
                    EXP
                      QVAR
                        QVARID
                          VARID
                            PsiElement(HaskellTokenType.varidRegexp)('i2n')
                      PsiElement(HaskellTokenType.()('(')
                      EXP
                        QVAR
                          QVARID
                            VARID
                              PsiElement(HaskellTokenType.varidRegexp)('crc')
                        QOP
                          QVAROP
                            QVARSYM
                              VARSYM
                                PsiElement(HaskellTokenType.Varsym)('.>>.')
                        PsiElement(HaskellTokenType.()('(')
                        EXP
                          QVAR
                            QVARID
                              VARID
                                PsiElement(HaskellTokenType.varidRegexp)('sz')
                          QOP
                            QVAROP
                              QVARSYM
                                VARSYM
                                  PsiElement(HaskellTokenType.-)('-')
                          PsiElement(HaskellTokenType.integertoken)('8')
                        PsiElement(HaskellTokenType.))(')')
                      PsiElement(HaskellTokenType.))(')')
                      QOP
                        QVAROP
                          QVARSYM
                            VARSYM
                              PsiElement(HaskellTokenType.Varsym)('.&.')
                      PsiElement(HaskellTokenType.integertoken)('0xFF')
                    PsiElement(HaskellTokenType.))(')')
                    QOP
                      QVAROP
                        PsiElement(HaskellTokenType.`)('`')
                        QVARID
                          VARID
                            PsiElement(HaskellTokenType.varidRegexp)('xor')
                        PsiElement(HaskellTokenType.`)('`')
                    QVAR
                      QVARID
                        VARID
                          PsiElement(HaskellTokenType.varidRegexp)('a')
                  PsiElement(HaskellTokenType.))(')')
                PsiElement(HaskellTokenType.))(')')
              PsiElement(HaskellTokenType.))(')')
              QOP
                QVAROP
                  PsiElement(HaskellTokenType.`)('`')
                  QVARID
                    VARID
                      PsiElement(HaskellTokenType.varidRegexp)('xor')
                  PsiElement(HaskellTokenType.`)('`')
              PsiElement(HaskellTokenType.()('(')
              EXP
                QVAR
                  QVARID
                    VARID
                      PsiElement(HaskellTokenType.varidRegexp)('crc')
                QOP
                  QVAROP
                    QVARSYM
                      VARSYM
                        PsiElement(HaskellTokenType.Varsym)('.<<.')
                PsiElement(HaskellTokenType.integertoken)('8')
              PsiElement(HaskellTokenType.))(')')
